train.py

```diff
@@ -216,10 +241,13 @@ def loss_fn(pred, labels):
         f"({gpu_mem_stats.max_reserved_pct:.2f}%)"
     )
 
+    if isinstance(model, FSDPModule) and parallel_dims.pp_enabled:
+        # reshard now to counteract an issue where FSDP's states got advanced during PP stage shape inference
+        model.reshard()
```

Comment:
@awgu following up, is this actually a 'bug' in FSDP or is this something we need to sort out at the trainer level (or inside DCP)? Should I file an issue to track?

Reply:
I think this is not a bug in FSDP per se but a bug in the composition of the PP shape inference and FSDP.

If PP shape inference runs a fake forward, then _any_ module that changes global state in the forward would have a similar issue.

Let me try to understand this better though. (sorry for the lack of understanding of the PP implementation)
- Is the PP stage shape inference orthogonal to the frontend tracer vs. manual splitting?
- It looks like we get `stage, model` --> apply TP/AC/FSDP --> build pipeline schedule. Is the shape inference happening in the "build pipeline schedule" part? (This seems to be the only way it makes sense that FSDP's states are getting advanced.)

Reply:
for manual stage (which is what i'm currently iterating on) we do shape inference during stage creation (__init__) today, but i'm proposing we move it to be lazy on first forward.  That's primarily so that we dont have to specify input shapes to each stage, for UX reasons, and so we also don't have to specify an input shape at all during stage creation (which is awkward to do any time before the first batch comes out of the dataloader).

for tracer, we need to specify the input shape for the model for the tracer to run, and that will always happen at stage creation time rather than later.

after applying FSDP, we should not actually be doing any fake-forward stuff today.

Reply:
In the **current** state of code, since you are feeding known input shapes to `ManualPipelineStage` constructor, why is there a need for shape inference? Or, did `ManualPipelineStage` just run it anyway? It seems to me that the easiest way to land the **current** state of code is make that shape inference conditional and move `model.reshard()` here. (Considering that developing lazy shape inference may take a while.)

Reply:
i need to double check what the reshard is for.

maybe previously, i was running shape-inference _and_ the manual stage creation was after fsdp wrap.

but now, i am doing stage creation before fsdp wrap, and I can’t use its shape inference there anyway because it would infer the wrong shapes at that early point.

Reply:
2D with DP ran fine locally when I removed this. its removed on the PR now, make sure CI is still happy..